applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp

```diff
@@ -599,25 +599,57 @@ size_t ComputeUpwindFactorCase(array_1d<double, 3>& rUpwindFactorOptions)
-bool CheckIfElementIsCutByDistance(const BoundedVector<double, NumNodes>& rNodalDistances)
+double ComputeUpwindFactorDerivativeWRTMachSquared(
+    const double localMachNumberSquared,
+    const ProcessInfo& rCurrentProcessInfo)
 {
-    // Initialize counters
-    unsigned int number_of_nodes_with_positive_distance = 0;
-    unsigned int number_of_nodes_with_negative_distance = 0;
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), section A.2
 
-    // Count how many element nodes are above and below the wake
-    for (unsigned int i = 0; i < rNodalDistances.size(); i++) {
-        if (rNodalDistances(i) < 0.0) {
-            number_of_nodes_with_negative_distance += 1;
-        }
-        else {
-            number_of_nodes_with_positive_distance += 1;
-        }
-    }
+    // read free stream values
+    // default CRITICAL_MACH - 0.99 and default UPWIND_FACTOR_CONSTANT - 1.0
+    const double critical_mach = rCurrentProcessInfo[CRITICAL_MACH];
+    const double upwind_factor_constant = rCurrentProcessInfo[UPWIND_FACTOR_CONSTANT];
 
-    // Elements with nodes above and below the wake are wake elements
-    return number_of_nodes_with_negative_distance > 0 &&
-           number_of_nodes_with_positive_distance > 0;
+    KRATOS_ERROR_IF(std::pow(localMachNumberSquared, 2.0) < std::numeric_limits<double>::epsilon())
+        << "ComputeUpwindFactorDerivativeWRTMachSquared: local mach number squared is less than zero." << std::endl;
+
+    return upwind_factor_constant * std::pow(critical_mach, 2.0) / std::pow(localMachNumberSquared, 2.0);
+}
+
+template <int Dim, int NumNodes>
+double ComputeUpwindFactorDerivativeWRTVelocitySquared(
+    const size_t upwindFactorCase, 
+    const array_1d<double, Dim>& rCurrentVelocity, 
+    const array_1d<double, Dim>& rUpwindVelocity,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), section A.2
+    const double current_element_mach_squared = ComputeLocalMachNumberSquared<Dim,NumNodes>(rCurrentVelocity, rCurrentProcessInfo);
+    const double upwind_element_mach_squared = ComputeLocalMachNumberSquared<Dim,NumNodes>(rUpwindVelocity, rCurrentProcessInfo);
+
+    if (upwindFactorCase == 0.0) {
+        return 0.0;
+    }
+    else if (upwindFactorCase == 1.0) {
+        const double upwind_factor_derivative = ComputeUpwindFactorDerivativeWRTMachSquared<Dim, NumNodes>(current_element_mach_squared,rCurrentProcessInfo);
+        const double mach_number_derivative = ComputeDerivativeLocalMachSquaredWRTVelocitySquared<Dim, NumNodes>(rCurrentVelocity, current_element_mach_squared,rCurrentProcessInfo);
+        return upwind_factor_derivative * mach_number_derivative;
```

Comment:
Since these lines are the same both for the supersonic accelerating case and for the supersonic decelerating case, would it make sense to make a shorter version of the function ComputeUpwindFactorDerivativeWRTVelocitySquared only with these three lines and then call it whenever you need it? Otherwise the upwindFactorCase gets checked in many functions and this will deteriorate your performande. I think it should be possible to make the implementation in such a manner that for each element you only check the upwindFactorCase once. I'm not sure whether I explained myself, we can discuss it tomorrow.

I would just do sth like:

```cpp
template <int Dim, int NumNodes>
double ComputeUpwindFactorDerivativeWRTVelocitySquared(
    const array_1d<double, Dim>& rVelocity, 
    const ProcessInfo& rCurrentProcessInfo)
{

    // Following Fully Simulataneous Coupling of the Full Potential Equation
    //           and the Integral Boundary Layer Equations in Three Dimensions
    //           by Brian Nishida (1996), section A.2
    const double mach_squared = ComputeLocalMachNumberSquared<Dim,NumNodes>(rVelocity, rCurrentProcessInfo);
    const double upwind_factor_derivative = ComputeUpwindFactorDerivativeWRTMachSquared<Dim, NumNodes>(mach_squared,rCurrentProcessInfo);
    const double mach_number_derivative = ComputeDerivativeLocalMachSquaredWRTVelocitySquared<Dim, NumNodes>(rVelocity, mach_squared, rCurrentProcessInfo);
    return upwind_factor_derivative * mach_number_derivative;

   
```